src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -48,38 +52,48 @@ static void uct_cuda_ipc_cache_purge(uct_cuda_ipc_cache_t *cache)
 {
     uct_cuda_ipc_cache_region_t *region, *tmp;
     ucs_list_link_t region_list;
+    int active;
+
+    UCT_CUDADRV_CTX_ACTIVE(active);
 
     ucs_list_head_init(&region_list);
     ucs_pgtable_purge(&cache->pgtable, uct_cuda_ipc_cache_region_collect_callback,
                       &region_list);
     ucs_list_for_each_safe(region, tmp, &region_list, list) {
-        UCT_CUDADRV_FUNC_LOG_ERR(
-                cuIpcCloseMemHandle((CUdeviceptr)region->mapped_addr));
+        if (active) {
+            UCT_CUDADRV_FUNC_LOG_ERR(
+                    cuIpcCloseMemHandle((CUdeviceptr)region->mapped_addr));
+        }
         ucs_free(region);
     }
     ucs_trace("%s: cuda ipc cache purged", cache->name);
 }
 
-static ucs_status_t uct_cuda_ipc_open_memhandle(CUipcMemHandle memh,
+static ucs_status_t uct_cuda_ipc_open_memhandle(uct_cuda_ipc_key_t *key,
                                                 CUdeviceptr *mapped_addr)
 {
+    ucs_status_t status = UCS_OK;
     const char *cu_err_str;
     CUresult cuerr;
 
-    cuerr = cuIpcOpenMemHandle(mapped_addr, memh,
+    ucs_recursive_spin_lock(&uct_cuda_ipc_map_lock);
+    cuerr = cuIpcOpenMemHandle(mapped_addr, key->ph,
                                CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS);
     if (cuerr != CUDA_SUCCESS) {
         if (cuerr == CUDA_ERROR_ALREADY_MAPPED) {
-            return UCS_ERR_ALREADY_EXISTS;
+            status = UCS_ERR_ALREADY_EXISTS;
+            goto out;
```

Comment:
need to flatten these if:
if (cuerr== SUCCESS || cuerr == ALREADY_EXISTS) {
 status == UCS_OK
}else{
 ucs_error()
 status = INVALID_PARAM
}

Reply:
unhandled

Reply:
@yosefe status is UCS_OK by default in this function; we cannot return UCS_OK when cuerr == ALREADY_EXISTS because functions which call `uct_cuda_ipc_open_memhandle` handle UCS_ERR_ALREADY_EXISTS differently from handling UCS_INVALID_PARAM (fatal error) which is different from handling UCS_OK. So I'm not sure how we can flatten to :
```C
if (cuerr== SUCCESS || cuerr == ALREADY_EXISTS) {
    status == UCS_OK;
} else {
    status == UCS_INVALID_PARAM;
}
```